### PR TITLE
Unittesting: freebsd changes, part 3

### PR DESCRIPTION
**Pull request by wjwithagen**
This requires Clang to be at least 3.7
It compiles all ceph code (expect the ones excluded in autogen_freebsd.sh)
It compiles all tests

Tests will run, but not all complete successfully,
  Currently the only one failing is osd-markup.sh

Other tests have been execluded from testing, see README.FreeBSD for that

Signed-off-by: Willem Jan Withagen <wjw@digiware.nl>

### CODE REVIEW

File: src/ceph-disk/tests/ceph-disk.sh [hunk 1]
```diff
@@ -54,7 +54,7 @@ CEPH_ARGS+=" --osd-failsafe-full-ratio=.99"
 CEPH_ARGS+=" --mon-host=$MONA"
 CEPH_ARGS+=" --log-file=$DIR/\$name.log"
 CEPH_ARGS+=" --pid-file=$DIR/\$name.pidfile"
-if test -d ../.libs ; then
+if [ -d ../.libs ]; then
```
Comment (liewegas) [inline src/ceph-disk/tests/ceph-disk.sh]: safer to stick with test, i think

**Comment by wjwithagen** [comment]
@tchaikov @liewegas 
I think this is enough for now.
This gets testing to 117 of 126.

There are a few that are easy onliners.
But the scripting in ceph-helpers.sh might be a bit more challenging, although teardown_error is not used yet. It will be in the commits to come.

File: src/test/cli/ceph-authtool/cap-bin.t [hunk 2]
```diff
@@ -2,5 +2,5 @@
   creating kring
 
   $ ceph-authtool --cap osd 'allow rx pool=swimming' kring
-  $ ceph-authtool kring --list|grep -P '^\tcaps '
+  $ ceph-authtool kring --list|grep -E '^\Wcaps '
```
Comment (tchaikov) [inline src/test/cli/ceph-authtool/cap-bin.t]: i don't understand why you'd prefer "\W" over "\t"?

Comment (wjwithagen) [inline src/test/cli/ceph-authtool/cap-bin.t]: @tchaikov 
-P is perl regex, and that is not in FreeBSD grep.
And `\t`is perl-speak.
So I try to stick to POSIX

Perhaps a bit of a braintwist at my end, but `\W` is abit more liberal in that it catches all types of white space. So it for whatever reason the output changes a bit, this will match that too.

**Comment by wjwithagen** [comment]
@tchaikov 
Anything I can do, other than wait for a qa run?

**Comment by tchaikov** [comment]
@wjwithagen i haven't got a chance to review all your changes. will look at them closer.

**Comment by wjwithagen** [comment]
@tchaikov 
Oke, take your time. Working on my sheet for Ceph day Cern anyways....

**Comment by wjwithagen** [comment]
@tchaikov 
Right, back in town...
So any suggestions are welcome

**Comment by tchaikov** [comment]
@wjwthagen will review your changeset next week.

**Comment by wjwithagen** [comment]
On 17-6-2016 12:41, Kefu Chai wrote:
> @wjwthagen will review your changeset next week.

Oke, no problem. Enjoy the weekend.

--WjW

**Comment by wjwithagen** [comment]
@tchaikov 
Hi Kefu,
I needed to rebase to get conflicts resolved, but now I'm getting the following error, in  tests/ceph-disk.sh.
So I have removed those commit from this pull:

Any suggestions what to fix??
```
tests/ceph-disk.sh:107: command_fixture:  local dir=testdir/test-ceph-disk
tests/ceph-disk.sh:108: command_fixture:  shift
tests/ceph-disk.sh:109: command_fixture:  local command=ceph-conf
tests/ceph-disk.sh:110: command_fixture:  shift
tttests/ceph-disk.sh:111: command_fixture:  which ceph-conf
ttests/ceph-disk.sh:111: command_fixture:  readlink -f /usr/srcs/Ceph/work.wip-wjw-freebsd-tests-2/ceph/src/ceph-conf
tests/ceph-disk.sh:111: command_fixture:  local fpath=/usr/srcs/Ceph/work.wip-wjw-freebsd-tests-2/ceph/src/ceph-conf
ttests/ceph-disk.sh:112: command_fixture:  readlink -f /usr/srcs/Ceph/work.wip-wjw-freebsd-tests-2/ceph/src/ceph-disk/ceph-conf
tests/ceph-disk.sh:112: command_fixture:  '[' x/usr/srcs/Ceph/work.wip-wjw-freebsd-tests-2/ceph/src/ceph-confx = xx ']'
tttests/ceph-disk.sh:112: command_fixture:  pwd
ttests/ceph-disk.sh:112: command_fixture:  readlink -f /usr/srcs/Ceph/work.wip-wjw-freebsd-tests-2/ceph/src/ceph-disk/ceph-conf
tests/ceph-disk.sh:112: command_fixture:  '[' x/usr/srcs/Ceph/work.wip-wjw-freebsd-tests-2/ceph/src/ceph-confx = xx ']'
tests/ceph-disk.sh:112: command_fixture:  return 1

```

File: qa/workunits/ceph-helpers.sh [hunk 3]
```diff
@@ -1452,13 +1472,13 @@ function main() {
...
-        code=0
+        teardown $dir || return 1
+        return 0
     else
         display_logs $dir
-        code=1
+        teardown_error $dir
+        return 1
     fi
-    teardown $dir || return 1
```
Comment (tchaikov) [inline qa/workunits/ceph-helpers.sh]: @wjwithagen can we leave this change for another changeset? this changes the behaviour of current tests.

File: qa/workunits/ceph-helpers.sh [hunk 4]
```diff
@@ -513,6 +524,9 @@ function destroy_osd() {
     local dir=$1
     local id=$2
 
+    # First mask osd down, otherwise it cannot be deleted later on
+    # is status has not propagated to monitor.
```
Comment (tchaikov) [inline qa/workunits/ceph-helpers.sh]: could you elaborate this a little bit? why we need to mark this osd down explicitly? and if we really need to do this. maybe you can move it down to line 531. just for consistency.

File: qa/workunits/ceph-helpers.sh [hunk 5]
```diff
@@ -1286,8 +1300,14 @@ function erasure_code_plugin_exists() {
     local plugin=$1
 
     local status
+    local grepstr
+    case x`uname`x in
```
Comment (tchaikov) [inline qa/workunits/ceph-helpers.sh]: no need to quote `uname` with "x". it is never empty in a sane OS.

File: qa/workunits/ceph-helpers.sh [hunk 6]
```diff
@@ -1286,8 +1300,14 @@ function erasure_code_plugin_exists() {
...
+        xLinuxx) grepstr="$plugin.*No such file" ;;
+        xFreeBSDx) grepstr="Cannot open.*$plugin" ;;
```
Comment (tchaikov) [inline qa/workunits/ceph-helpers.sh]: maybe we can have a
```
*)
```
to catch all other cases?

File: src/ceph-disk/tests/test_main.py [hunk 7]
```diff
@@ -310,6 +311,10 @@ def get_partition_type(dev):
             assert expect == main.list_devices()
 
     def test_list_dmcrypt_data(self):
+        if platform.system() == "FreeBSD":
```
Comment (tchaikov) [inline src/ceph-disk/tests/test_main.py]: @dachary does this change look good to you?

File: src/ceph-disk/tests/test_main.py [hunk 8]
```diff
@@ -14,6 +14,7 @@
 #
 from mock import patch, DEFAULT
 import os
+import platform
```
Comment (tchaikov) [inline src/ceph-disk/tests/test_main.py]: @wjwithagen nit, the commit message could read "FreeBSD does not have blkid"

File: src/vstart.sh [hunk 9]
```diff
@@ -92,7 +92,11 @@ start_rgw=0
 ip=""
 nodaemon=0
 smallmds=0
-short=0
+if [ x`uname`x = xFreeBSDx ] ; then
```
Comment (tchaikov) [inline src/vstart.sh]: no need to quote `uname` with "x".

File: src/global/pidfile.cc [hunk 10]
```diff
@@ -162,7 +162,12 @@ int pidfh::open(const md_config_t *conf)
   pf_dev = st.st_dev;
   pf_ino = st.st_ino;
 
-  struct flock l = { F_WRLCK, SEEK_SET, 0, 0, 0 };
+  struct flock l;
+  memset(&l, 0, sizeof(l));
+  l.l_type = F_WRLCK;
+  l.l_whence = SEEK_SET;
+  l.l_start = 0;
+  l.l_len = 0;
```
Comment (tchaikov) [inline src/global/pidfile.cc]: could put

```c++
  struct flock l = {
    .l_type = F_WRLCK,
    .l_whence = SEEK_SET,
    .l_start = 0,
    .l_len = 0
  };
```
instead.

File: src/vstart.sh (hunk 9 above)
Comment (tchaikov) [inline src/vstart.sh]: but zfs has longer xattr limits, i assume?

File: src/ceph-disk/ceph_disk/main.py [hunk 11]
```diff
@@ -743,7 +743,11 @@ def is_mounted(dev):
     Check if the given device is mounted.
     """
     dev = os.path.realpath(dev)
-    with file('/proc/mounts', 'rb') as proc_mounts:
+    proc_prefix = ''
+    if platform.system() == "FreeBSD":
+        # Use the Linux compatability tree
```
Comment (tchaikov) [inline src/ceph-disk/ceph_disk/main.py]: s/compatability/compatibility/

File: src/common/config_opts.h [hunk 12]
```diff
@@ -1078,7 +1096,11 @@ OPTION(filestore_max_sync_interval, OPT_DOUBLE, 5)    // seconds
 OPTION(filestore_min_sync_interval, OPT_DOUBLE, .01)  // seconds
 OPTION(filestore_btrfs_snap, OPT_BOOL, true)
 OPTION(filestore_btrfs_clone_range, OPT_BOOL, true)
+#if defined(__linux__)
```
Comment (tchaikov) [inline src/common/config_opts.h]: maybe we can enable `filestore_zfs_snap` by default? as zfs is included by some major linux distros recently. @ukernel what do you think?

File: src/common/xattr.h [hunk 13]
```diff
@@ -19,12 +19,14 @@
 extern "C" {
 #endif
 
+#if defined(__linux__)
```
Comment (tchaikov) [inline src/common/xattr.h]: i am fine with this change, but it's always good to have `ENOATTR` defined, right? or we need to check if `ENODATA` is defined also here? to avoid cyclic definition? as we have
```
#define ENODATA ENOATTR
```
in `compat.h`.

[OT] if `xattr.c` is the only consumer of `ENOATTR`, maybe we can move this definition into `xattr.c` in another commit.

File: qa/workunits/ceph-helpers.sh (hunk 4 above)
Comment (wjwithagen) [inline qa/workunits/ceph-helpers.sh]: I only added the ` [ x`uname`x = xFreeBSDx ] || send_signal=0`.
The part about the OSD maks is not mine. Perhaps a left over from rebasing earlier on.
I will remove that piece of comments.

But this is the wrong place... __teardown_btrfs will not be called on a FreeBSD platform.

File: qa/workunits/ceph-helpers.sh (hunk 5 above)
Comment (wjwithagen) [inline qa/workunits/ceph-helpers.sh]: Sane would be relative in this crazy world. 8D
I've seen too many things come up empty... The last one being readlink(1).
But you are right chances are slim.

And I'll put in a wildcard as default catch

File: qa/workunits/ceph-helpers.sh (hunk 3 above)
Comment (wjwithagen) [inline qa/workunits/ceph-helpers.sh]: @tchaikov 
It is not my intention to change any of the behaviour of the tests.
Other than:
  - Fix FreeBSD issues
  - Keep data around when something fails, so one can look at the logfiles
So if it does change standard behaviour on Linux, when tests go as planned, that should be a mistake.

But I'm fine with taking this change to a next set of commits for the next pull.
Note that I talked about it with a few people at Ceph day, and it would be nice if tests also start taking care of situations where things actually fail, and people doing testing/debugging/porting would have easy access to more information.

File: src/ceph-disk/tests/test_main.py (hunk 8 above)
Comment (wjwithagen) [inline src/ceph-disk/tests/test_main.py]: Will fix

File: src/ceph-disk/tests/test_main.py (hunk 7 above)
Comment (wjwithagen) [inline src/ceph-disk/tests/test_main.py]: @tchaikov @dachary 
FreeBSD does not have any of the dm stuff at all. The short cut was to disable here.
As I did at a few other places

File: src/common/config_opts.h (hunk 12 above)
Comment (wjwithagen) [inline src/common/config_opts.h]: @tchaikov @ukernel 
I have not done so, because I would liek to first see that things are working with ZFS as expected.
And then actually start integrating it fully.
Then the options would also become real zfs options instead of:
    `filestore_*_xattr_*_other`
For now I'm just overulling the default on FreeBSD, otherwise there was going to be another lage set of patches for that. But I'll do that in a set of dedicated ZFS patches.

File: src/common/xattr.h (hunk 13 above)
Comment (wjwithagen) [inline src/common/xattr.h]: The definition was already there, I just wrapped it in ifdef because might case redefinition issues, and compiler warnings... 
But I think ENOATTR is not used in ANY of the code other than compat.h.
And there it is to create compatibility with FreeBSD.
So yes, I think this whole piece could/should go.??

File: src/global/pidfile.cc (hunk 10 above)
Comment (wjwithagen) [inline src/global/pidfile.cc]: Oke, that is the nicer style. Are all not mentioned values than set to 0?
Because that is what the memset() does, althoug I suspect that that is not required for the call we're doing

File: src/vstart.sh (hunk 9 above)
Comment (wjwithagen) [inline src/vstart.sh]: Fixed the xx's
And yes ZFS has long attributes, but Sage commented during Ceph day that we should use --short for some of the tests. And this is just to make sure that we do that when running FreeBSD, because is is known to sometimes fail the test in ceph-osd. This prevents that

File: qa/workunits/ceph-helpers.sh [hunk 14]
```diff
@@ -17,6 +17,15 @@
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 # GNU Library Public License for more details.
 #
+
+# to be able to run standalone define some VARIABLES
```
Comment (tchaikov) [inline qa/workunits/ceph-helpers.sh]: this is not related FreeBSD port. could you remove it from this commit?

File: qa/workunits/ceph-helpers.sh [hunk 15]
```diff
@@ -123,6 +132,16 @@ function teardown() {
     rm -fr $dir
 }
 
+function teardown_error() {
```
Comment (tchaikov) [inline qa/workunits/ceph-helpers.sh]: this function is not called, could you please remove it also?

File: src/common/xattr.h (hunk 13 above)
Comment (tchaikov) [inline src/common/xattr.h]: i am inclined to keep it as it is, unless there is case that we see redefinition issues or warnings. why we could have redefinition issues because of this `#define`?

i don't think the whole piece should go. see 886f5a91f35d04cd4a721f1a66a2ef4a2c39fd5e

Comment (wjwithagen) [inline src/common/xattr.h]: @tchaikov 
I'll put it back in.

But in that case I would be very mich in favour of reversing the whole deal...
Rip out all references to ENODATA and actually start using ENOATTR every where.
Under Linux we then can declare ENODATA = ENOATTR.
But then again, that is perhaps for another set of commits.

Comment (wjwithagen) [inline src/common/xattr.h]: @tchaikov 
(forget my previous message)

I looked at it again, and it is foolish to put in for anything else than Linux.
If this case would arise under FreeBSD (or other  OS)  I'd rather have it signal an error because at this point we are very likely going to patch up ENOATTR with the wrong value.
So rather than doing that, I is better for the compiler to tell that ENOATTR is not defined. 
The way to fix that is to include include/compat.h.

If you prefer I can add a compiler `#WARNING` to that extend in the `#else` clause.

File: qa/workunits/ceph-helpers.sh (hunk 15 above)
Comment (wjwithagen) [inline qa/workunits/ceph-helpers.sh]: @tchaikov 
This is in preparation that we start doing some sensible things when a test actually fails. We then do need the poor remains of the carcass to see what happened. But that can go into another commit/pull as well.
